### PR TITLE
Fix model viewer

### PR DESCRIPTION
**Pull request by chsterz**
Several fixes for the model viewer.
Improves usability and stability of the model inspector.

### CODE REVIEW

File: plugins/modelinspector/modelcellmodel.cpp [hunk 1]
```diff
@@ -49,7 +49,9 @@ static bool sourceIsQQmlListModel(const QAbstractItemModel *model)
 
 void ModelCellModel::setModelIndex(const QModelIndex &idx)
 {
+
```
Comment (krf) [inline plugins/modelinspector/modelcellmodel.cpp]: Style: Redundant newlines here & below

File: plugins/modelinspector/modelinspector.cpp [hunk 2]
```diff
@@ -156,14 +156,15 @@ void ModelInspector::cellSelectionChanged(const QItemSelection &selection)
     QModelIndex index;
     if (selection.size() >= 1)
         index = selection.at(0).topLeft();
+
+    index = m_modelContentProxyModel->mapToSource(index);
```
Comment (krf) [inline plugins/modelinspector/modelinspector.cpp]: Rather call that one `sourceIndex`

File: plugins/qmlsupport/qmllistpropertyadaptor.cpp [hunk 3]
```diff
@@ -50,6 +50,8 @@ int QmlListPropertyAdaptor::count() const
         = reinterpret_cast<QQmlListProperty<QObject> *>(const_cast<void *>(var.data()));
     if (!prop)
         return 0;
+    if (prop->count == nullptr)
```
Comment (krf) [inline plugins/qmlsupport/qmllistpropertyadaptor.cpp]: Style: `!prop->count` (here & below)

File: plugins/modelinspector/modelcontentproxymodel.h [hunk 4]
```diff
@@ -50,7 +50,8 @@ class ModelContentProxyModel : public QIdentityProxyModel
 public:
     enum Roles {
         DisabledRole = GammaRay::UserRole + 1,
-        SelectedRole
+        SelectedRole,
+        EmptyNameRole
```
Comment (krf) [inline plugins/modelinspector/modelcontentproxymodel.h]: I'd name that a bit more descriptive, e.g. `IsDisplayStringEmptyRole`

=> You'll know it checks the display role string
=> You'll know it'll return a bool

File: plugins/modelinspector/modelcontentdelegate.cpp [hunk 5]
```diff
@@ -64,6 +64,12 @@ void ModelContentDelegate::paint(QPainter *painter, const QStyleOptionViewItem &
         // TODO also render selection for currently selected cells
     }
 
+    if (index.data(ModelContentProxyModel::EmptyNameRole).toBool()) {
+        option.palette.setColor(QPalette::Text, QPalette::Disabled);
+        option.text = tr("< unnamed at row:" + QString::number(index.row())
+                              + " column:" + QString::number(index.column()) + " >");
```
Comment (vkrause) [inline plugins/modelinspector/modelcontentdelegate.cpp]: This is neither translatable nor does it compile. Would need to be something like tr("<unnamed row %1 column %2>").arg(index.row()).arg(index.column())

**Comment by chsterz** [comment]
fixed and rebased – based on comments.

File: plugins/qmlsupport/qmllistpropertyadaptor.cpp [hunk 6]
```diff
@@ -60,9 +62,13 @@ PropertyData QmlListPropertyAdaptor::propertyData(int index) const
     auto var = object().variant(); // we need to keep that alive for the runtime of this method
     QQmlListProperty<QObject> *prop
         = reinterpret_cast<QQmlListProperty<QObject> *>(const_cast<void *>(var.data()));
+    if (!prop->count)
+        return pd;
     if (!prop || index >= prop->count(prop))
```
Comment (krf) [inline plugins/qmlsupport/qmllistpropertyadaptor.cpp]: Checking for `prod != nullptr` after dereferencing it (see line 65) makes no sense. You need to reorder those statements.

```
if (!prop)
    return pd;

if (!prop->count || index >= ...)
    return pd;
```

File: plugins/qmlsupport/qmllistpropertyadaptor.cpp [hunk 7]
```diff
@@ -60,9 +62,13 @@ PropertyData QmlListPropertyAdaptor::propertyData(int index) const
...
         return pd;
 
+    if (!prop->at)
```
Comment (krf) [inline plugins/qmlsupport/qmllistpropertyadaptor.cpp]: Doing this first (after checking for `prod != nullptr`) makes this code path faster. `prop->at` is just a pointer comparison, while the check above may call a (virtual) function.

File: plugins/modelinspector/modelcontentdelegate.cpp [hunk 8]
```diff
@@ -64,6 +64,11 @@ void ModelContentDelegate::paint(QPainter *painter, const QStyleOptionViewItem &
         // TODO also render selection for currently selected cells
     }
 
+    if (index.data(ModelContentProxyModel::IsDisplayStringEmptyRole).toBool()) {
+        option.palette.setColor(QPalette::Text, option.palette.color(QPalette::Disabled, QPalette::Text));
+        option.text = tr("<unnamed row %1 column %2>").arg(index.row()).arg(index.column());
```
Comment (krf) [inline plugins/modelinspector/modelcontentdelegate.cpp]: Better: `<unnamed: row %1, column %2>`

File: plugins/modelinspector/modelinspector.cpp [hunk 9]
```diff
@@ -156,20 +156,22 @@ void ModelInspector::cellSelectionChanged(const QItemSelection &selection)
     QModelIndex index;
     if (selection.size() >= 1)
         index = selection.at(0).topLeft();
-    m_cellModel->setModelIndex(index);
 
-    if (!index.isValid()) {
+    QModelIndex sourceIndex;
```
Comment (krf) [inline plugins/modelinspector/modelinspector.cpp]: Sorry, another one: Could merge with next line, constify

**Comment by vkrause** [comment]
This probably should also go into 2.7 now.

**Pull request closed without merging**